### PR TITLE
pull images before upgrading

### PR DESCRIPTION
**Pull request by deepikasl**
https://github.com/Shippable/base/issues/882

Cleared `"services": [ ]` in state.json. After running base installer, `services` got populated with all env's of each service.

### CODE REVIEW

File: scripts/bootstrapApp.sh [hunk 1]
```diff
@@ -610,13 +610,396 @@ update_service_list() {
...
 main() {
   __process_marker "Updating system config"
   generate_serviceuser_token
   generate_root_bucket_name
 
   local is_upgrade=$(cat $STATE_FILE | jq -r '.isUpgrade')
   if [ "$INSTALL_MODE" == "production" ]; then
+    update_service_list
+    load_services
+    provision_services
```
Comment (shrivara) [inline scripts/bootstrapApp.sh]: This should not happen here. I think the discussion was to only update the service list in `state.json` and pull images here. The provisioning still needs to happen in `provisionServices.sh`.

Comment (shrivara) [inline scripts/bootstrapApp.sh]: We also need to move `update_service_list` to be before this as that generates a list of services.

Comment (shrivara) [inline scripts/bootstrapApp.sh]: When you are testing this, clear out the `services[]` array in `state.json` and run a fresh install. After the install, it should have all the services based on the `masterIntegrations` enabled.

File: scripts/bootstrapApp.sh [hunk 2]
```diff
@@ -459,12 +459,6 @@ provision_api_local() {
 
   sudo docker rm -f api || true
 
-  # Pull image before attempting to run to ensure the image is always updated in
-  # case it was overwritten. Like in case of :latest.
-  local pull_api_cmd="sudo docker pull $image"
```
Comment (shrivara) [inline scripts/bootstrapApp.sh]: Why was this removed?

File: scripts/provisionServices.sh [hunk 3]
```diff
@@ -381,7 +196,6 @@ __run_service() {
 
 provision_www() {
   local sleep_time=30
-  __save_service_config www " --publish 50001:50001/tcp" "--mode global --name www --network ingress --with-registry-auth --endpoint-mode vip"
```
Comment (shrivara) [inline scripts/provisionServices.sh]: If we remove this from here, where does this happen?

Comment (deepikasl) [inline scripts/provisionServices.sh]: This is happening [here](https://github.com/Shippable/base/pull/883/files#diff-74d1098cdf460b20247fa97b07e31322R881)

File: scripts/bootstrapApp.sh [hunk 4]
```diff
@@ -637,9 +914,10 @@ main() {
     update_dynamic_nodes_integration_id
     generate_system_config
     create_system_config
-    update_service_list
     restart_api
   else
+    update_service_list
+    pull_images
```
Comment (shrivara) [inline scripts/bootstrapApp.sh]: We can remove this as we won't have `services` machines on local, so it will never pull images.

Comment (deepikasl) [inline scripts/bootstrapApp.sh]: okay.

File: scripts/provisionServices.sh (hunk 3 above)
Comment (shrivara) [inline scripts/provisionServices.sh]: That's not publishing the ports, is it?

File: scripts/provisionServices.sh [hunk 5]
```diff
@@ -393,10 +369,8 @@ provision_services() {
     local service=$(echo $services | jq -r '.['"$i-1"'] | .name')
     local provisioned_service=$(echo $provisioned_services | jq -r '.[] | select (.=="'$service'")')
     if [ -z "$provisioned_service" ]; then
-      __save_service_config $service "" " --name $service --network ingress --with-registry-auth --endpoint-mode vip" $service
       local service_image=$(cat $STATE_FILE \
```
Comment (shrivara) [inline scripts/provisionServices.sh]: This can be removed now.
